src/main/java/org/folio/rest/impl/BLUsersAPI.java

```diff
@@ -816,27 +829,57 @@ public void postBlUsersLogin(boolean expandPerms, List<String> include, String u
           .ifPresent(header -> headers.put(HttpHeaders.USER_AGENT, header));
         Optional.ofNullable(xForwardedFor)
           .ifPresent(header -> headers.put(X_FORWARDED_FOR_HEADER, header));
-        loginResponse[0] = client.request(HttpMethod.POST, entity, moduleURL, headers);
+
+        List<String> finalInclude = include;
+
+        clientForLogin.request(HttpMethod.POST, entity, moduleURL, headers)
+          .thenAccept(loginResponse -> {
+            //then get user by username, inject okapi headers from the login response into the user request
+            //see 'true' flag passed into the chainedRequest
+            handleResponse(loginResponse, false, false, true, aRequestHasFailed, asyncResultHandler);
+
+            String token = loginResponse.getHeaders().get(OKAPI_TOKEN_HEADER);
+            String tenant = getTenant(token);
+            okapiHeaders.put(OKAPI_TENANT_HEADER, tenant);
+            HttpClientInterface client = HttpClientFactory.getHttpClient(okapiURL, tenant);
+
+            try {
+              getUserWithPerms(expandPerms, okapiHeaders, asyncResultHandler, userUrl, finalInclude, token, tenant, client);
+            } catch (Exception e) { //NOSONAR
+              client.closeClient();
+              asyncResultHandler.handle(io.vertx.core.Future.succeededFuture(
+                PostBlUsersLoginResponse.respond500WithTextPlain(e.getLocalizedMessage())));
+            } finally {
+              clientForLogin.closeClient();
+            }
+          })
+          .exceptionally(throwable -> {
+            clientForLogin.closeClient();
+            asyncResultHandler.handle(io.vertx.core.Future.succeededFuture(
+              PostBlUsersLoginResponse.respond500WithTextPlain(throwable.getLocalizedMessage())));
+            return null;
+          });
       } catch (Exception ex) {
-        client.closeClient();
+        clientForLogin.closeClient();
         asyncResultHandler.handle(io.vertx.core.Future.succeededFuture(
           PostBlUsersLoginResponse.respond500WithTextPlain(ex.getLocalizedMessage())));
-        return;
       }
-      //then get user by username, inject okapi headers from the login response into the user request
-      //see 'true' flag passed into the chainedRequest
-      userResponse[0] = loginResponse[0].thenCompose(client.chainedRequest(
-        userUrl, okapiHeaders, false, null, handlePreviousResponse(false,
-          false, true, aRequestHasFailed, asyncResultHandler)));
-
-      //populate composite based on includes
-      int includeCount = include.size();
+    }
+  }
+
+  private void getUserWithPerms(boolean expandPerms, Map<String, String> okapiHeaders, Handler<AsyncResult<javax.ws.rs.core.Response>> asyncResultHandler,
```

Comment:
I support the ideaof breaking something out into its own method. But I question whether this alone really helps reduce complexity of the parent method. Is that the purpose here, to reduce complexity? I would suggest leaving this kind of refactoring for another day when we can make more of a difference.

Reply:
Ticket created: https://issues.folio.org/browse/MODUSERBL-170